### ISSUE TITLE
[question] testing with normalization env wrapper

### ISSUE THREAD

**Issue by ggenesum** [hill-a/stable-baselines#1077]
I am trying to use PPO2 in a gym trading environnent. 
I have two environnements for test and train, that simulate trading scenarios from histrorical data at different periods.
Reward is prices@t+1/prices@t with some random.

```
train_envs =  SubprocVecEnv([lambda:  TradingVecEnv(df_train) for i in range(n_cpu)])
test_envs =  SubprocVecEnv([lambda:  TradingVecEnv(df_test) for i in range(n_cpu)])
train_envs = VecNormalize(train_envs, norm_obs=True, norm_reward=True,
               clip_obs=max_grad_norm, gamma=gamma_norm) 
test_envs =  VecNormalize(test_envs, norm_obs=True, norm_reward=False,
       clip_obs=max_grad_norm, gamma=gamma_norm) 
```
It seems the model is overfitting. It shows no progress on test environment.
It is very well possible and I am currently tuning some hyperparameters including network architecture (on a custom MlpLnLstmPolicy). Also I need more data (and probably a better computer)

But I'm surprised to see no progress at all on test set while having such smooth learning curves.
It is either totally overfitting, or something is wrong with train reward logging on tensorboard.
Or it is when testing with the other environment.

It started working on train set when I added the VecNormalize wrapper with reward normalization.

I would like to make sure I use VecNormalize the right way.
I defined a callback function wich has the test env as self.eval_env

```
def _on_step(self) -> bool:
    if self.n_calls == 1:
        self.start_time = time.clock()
        self.elapsed_time = 0
    if self.eval_freq > 0 and self.n_calls % self.eval_freq == 0:
        self.elapsed_time += time.clock() - self.start_time
        if self.elapsed_time > 2500:
            self.elapsed_time = 2500

        sync_envs_normalization(self.training_env, self.eval_env)

        #model.set_env(self.eval_env)
        dones , states = [False for i in range(self.eval_env.num_envs)], None
        rwds = []
        obs = self.eval_env.reset()
        for i in range(1000):
            actions, states = model.predict(obs, state=states, deterministic=False)
            obs, rewards, dones, info = self.eval_env.step(actions)
            rwds = np.concatenate((rewards, rwds), axis=0)
        reward = np.mean(rwds)
        if math.isnan(reward) or reward == None:
            reward = 0.0
        trial.report(reward, self.elapsed_time)
        if trial.should_prune():
            raise optuna.TrialPruned()
        if self.verbose > 0:
            print("Eval num_timesteps={}, "
                  "reward={:.6f}".format(self.num_timesteps, reward))
        if self.elapsed_time >= 2500:
            return False
        if reward == 0.0:
            return False
        self.start_time = time.clock()
    return True
```

I have read the code of sync_envs_normalization but I am still unsure of what it is doing exactly.
Also is batch size limited to the real number of CPU cores with SubProcVecEnv ?
Maybe increasing could help prevent overfitting.

Thanks for any insight

**Comment by Miffyli** [comment]
On a quick glimpse the evaluation code seems right, however I am 100% certain how things play out with the parallelized evaluation (it _might_ bias towards shorter episodes, if there are any). I recommend using the [pre-existing](https://github.com/hill-a/stable-baselines/blob/master/stable_baselines/common/callbacks.py#L295) evaluation code just to make sure it is not code's fault.

> Also is batch size limited to the real number of CPU cores with SubProcVecEnv ?
No, you can use as many environments and/or as large batches as you want.

> But I'm surprised to see no progress at all on test set while having such smooth learning curves.
This is very common to reinforcement learning: it overfits to the training environments, and without specific tinkering rarely extends beyond the training env. See e.g. [this paper on procgen](https://arxiv.org/abs/1912.01588). I would not be surprised if for a fixed datasets and such things get even worse.

**Comment by ggenesum** [comment]
I think the evaluate_policy function works only with a single env, but after training on 10 envs with PPO2 (or maybe it is because using a recurrent policy), i am unable to test with a single env.
Thanks  I will have a look on this paper.

**Comment by Miffyli** [comment]
> or maybe it is because using a recurrent policy

Yes, this is the reason (see  [note](https://stable-baselines.readthedocs.io/en/master/guide/examples.html#recurrent-policies) on docs) :)

**Comment by ggenesum** [comment]
Ok i See. After few more test I think the test callback is working as expected, I will update  if I find something related to stable-baselines. Thanks
